### ISSUE TITLE
[BUG] SaltReqTimeoutError / master could not be contacted / Multi-Master

### ISSUE THREAD

**Issue by erespaulo** [saltstack/salt#58028]
**Description**

Hello guys!

I am looking for a solution to a communication problem between our Master -> Syndic -> Minions structure.

We currently have 3 masters, 11 syndics and approximately 6K minions. This structure is still small compared to larger uses of SaltStack's potential, however we are faced with several moments with Timeouts and generating low communication with Minions (We have moments that only 600 minions respond).

#What is your environment? Production integrations
#What steps will reproduce the issue? 12 hours or so, the problem happens.
#What Operating System? Centos 7
#What would you expect to be the outcome? Normal Operation.

I have already checked several bug's reported in Git since much older versions and unfortunately this problem does not appear to be related to a version bug, but I believe that it would be already fixed.

#Related Issues : 43436 ; 55629 ; 53147.

I have already performed the "salt-run state.event pretty = True", 0mq Tunning attempts and have not found a root cause that may be affecting the master.

I did not find any reference to these problems being dealt with in the release notes superior to our Master, which may point out that it is not a real Bug.

**Setup**

-  salt-minions versions ranged from 2015 to 3000.2

- salt --versions-report
Salt Version:
           Salt: 2019.2.5
Dependency Versions:
           cffi: Not Installed
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: 1.2.5
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr  2 2020, 13:16:51)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
System Versions:
           dist: centos 7.8.2003 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1062.12.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.8.2003 Core

-  salt master has following HW resource.

16 core CPUs
64G RAM  and around 8G used

**master.conf**

> yaml_utf8: True
worker_threads: 32
show_jid: True
file_roots:
  base:
    - /srv/salt
pillar_roots:
  base:
    - /srv/pillar
\# Timeout
key_cache: 'sched'
timeout: 30
gather_job_timeout: 60
\# Allow cp.push from minions
file_recv: True
order_masters: True
sock_pool_size: 15
max_open_files: 100000
log_level_logfile: all
\#Tunning ZMQ Settings - 01072020
\#salt_event_pub_hwm: 256000
\#event_publisher_pub_hwm: 128000
\#http://api.zeromq.org/3-2:zmq-setsockopt - 60 a 70%
pub_hwm: 120000
zmq_backlog: 130000
con_cache: True
file_buffer_size: 16777216
max_event_size: 10485760
reactor_worker_hwm: 100000
reactor_worker_threads: 16
reactor_refresh_interval: 3600
tcp_keepalive_idle: -1

**redis.conf**

> redis.host: <SERVER IP>
redis.port: <SERVER PORT>
redis.db: 0
cache: redis
cache.redis.host: <SERVER IP>
cache.redis.port: <SERVER PORT>
cache.redis.db: 0
minion_data_cache: True
master_job_cache: redis
job_cache_store_endtime: True
\# Keep jobs in redis in hours
keep_jobs: 24

**api.conf**

> \######          API  Settings         #####
> \###########################################
rest_cherrypy:
  port: 9080
  host: 0.0.0.0
  collect_stats: false
  disable_ssl: true
  log_level_logfile: all
  log_access_file: /var/log/salt/api_access.log
  log_error_file: /var/log/salt/api_error.log
  rest_timeout: 120

- Python Minions
Park with some machines in Py2 (in the process of updating) and Py3.

**In the Syndics i have the optional installation of Salt Minion looking for performance.**

> ipv6: False
minion_id_lowercase: True
random_startup_delay: 10
acceptance_wait_time: 60
random_reauth_delay: 300
recon_default: 5000
recon_max: 120000
recon_randomize: True
master_tops_first: True
grains_cache: True
grains_refresh_every: 1440
grains_deep_merge: True
#log_level_logfile: debug
log_level_logfile: info
log_rotate_max_bytes: 10000000
log_rotate_backup_count: 5
zmq_monitor: True
master_alive_interval: 30
master_shuffle: True

**Additional context**

#1 - Syndic Logs

2020-07-24 18:04:59,577 [salt.transport.zeromq:1244][DEBUG   ][29491] SaltReqTimeoutError, retrying. (2/3)
2020-07-24 18:04:59,578 [salt.transport.zeromq:1244][DEBUG   ][29491] SaltReqTimeoutError, retrying. (2/3)
2020-07-24 18:04:59,578 [salt.transport.zeromq:1244][DEBUG   ][29491] SaltReqTimeoutError, retrying. (2/3)
2020-07-24 18:04:59,579 [salt.transport.zeromq:229 ][DEBUG   ][29491] This is not the last AsyncZeroMQReqChannel instance. Not closing yet.
2020-07-24 18:04:59,579 [salt.minion      :1462][INFO    ][29491] fire_master failed: master could not be contacted. Request timed out.
2020-07-24 18:04:59,580 [salt.transport.zeromq:1214][DEBUG   ][29491] Re-init ZMQ socket: Message timed out
2020-07-24 18:04:59,580 [salt.transport.zeromq:1189][DEBUG   ][29491] Trying to connect to: tcp://<SERVERIP>:4506

#2 - Other Sample.

2020-07-27 11:01:46,781 [salt.minion      :2026][INFO    ][30212] Returning information for job: 20200727110001449196
2020-07-27 11:01:46,781 [salt.minion      :2026][INFO    ][30212] Returning information for job: 20200727110001449196
2020-07-27 11:01:46,804 [salt.minion      :1462][INFO    ][30212] fire_master failed: master could not be contacted. Request timed out.
2020-07-27 11:01:47,652 [salt.minion      :1462][INFO    ][30212] fire_master failed: master could not be contacted. Request timed out.
2020-07-27 11:01:47,654 [salt.minion      :1462][INFO    ][30212] fire_master failed: master could not be contacted. Request timed out.
2020-07-27 11:01:48,649 [salt.minion      :2026][INFO    ][30212] Returning information for job: 20200727110146303385
2020-07-27 11:01:48,654 [salt.minion      :1462][INFO    ][30212] fire_master failed: master could not be contacted. Request timed out.
2020-07-27 11:01:48,657 [salt.minion      :1462][INFO    ][30212] fire_master failed: master could not be contacted. Request timed out.

We have a continuous restart script as a palliative solution.
Could you support us to find what could be done in my environment?

**Comment by garethgreenaway** [comment]
@erespaulo Thanks for the report.  To confirm, are some of your minions running one of the 2015.x releases of Salt?

**Comment by erespaulo** [comment]
Hi @garethgreenaway  , I have 12 minions (2015.8.10) and 1 (2015.8.8) , the most bigger part is 2019.2.5 .

**Comment by erespaulo** [comment]
@garethgreenaway , My problem happens to version 2019.2.5 . My park is almost all in 2019.2.5 .

**Comment by erespaulo** [comment]
@sagetherage  , I was following this item and we also have a Multi master structure, Multi Syndic. If that helps in any way.

https://github.com/saltstack/salt/issues/18751

**Comment by sagetherage** [comment]
@erespaulo 2019.2.5 is only in CVE support, are you able to upgrade to a newer version of Salt that is supported?

**Comment by erespaulo** [comment]
@sagetherage , I am conducting a project to evaluate the situation on the version 3000.3 with master's and minions.

Unfortunately there is a time of testing and process for this to be transposed to the current structure, for this reason I would like to exhaust the possibilities of correction still in version 2019.2.

**Comment by sagetherage** [comment]
@Ch3LL

**Comment by Ch3LL** [comment]
What is your hardware usage like? RAM, CPU? 
Are there possibly any memory leaks occurring?

**Comment by erespaulo** [comment]
hi @Ch3LL , our Masters has 64Gb RAM , 16 virtual cores. 

In System / Messages Logs there's no evidence of problemas with memory.
I am still working in some testing labs using minionswarm to try to find out something.

**Comment by Ch3LL** [comment]
k let me know when you want to keep troubleshooting after your tests

**Comment by erespaulo** [comment]
@Ch3LL , 

I did a benchmark test for a lab structure using version 3000.3 for all Masters and Minions.

For minions, I used the minionswarm.py script.

In this scenario I used 1 MASTER (16 proc, 64RAM), 8 servers (2 servers with 16 proc, 64 RAM and 6 wit 8 proc,16GB RAM) with the minions script and 1 Server for Redis / Elastic (Cache).

During the execution of basic tests such as the use of "test.ping", noticed the memory consumption on the servers during the processing (around 10% of the minion consumption) of this information and as soon as the resource is processed, it goes back to the server.

As a result of the test, I made a FOR loop to stress the environment and we verified that after a certain time a degradation of salt-minions is detectable.

Scenario: 1 master, 8 minion injectors
Test with 300 minions (Server5) = Ok
Test with 1400 minions (Server5, Server7) = Ok
Test with 1900 minions (Server5, Server7, Server8) = OK
Test with 2200 minions (Server5, Server7, Server8, Server1) = Ok
Test with 2500 minions (Server5, Server7, Server8, Server1, Server2) = Ok
Test with 2800 minions (Server5, Server7, Server8, Server1, Server2, Server3) = Ok
Test with 3100 minions (Server5, Server7, Server8, Server1, Server2, Server3, Server4) = Ok
Test with 3400 minions (Server5, Server7, Server8, Server1, Server2, Server3, Server4, Server6) = Ok

The RAM consumption of the MASTER server with 3400 minions was 5GB from the master.

We started testing at 10:56 AM and in growing use of minions, where it remained stable until 3:16 PM, where Server5 degradation started.

The conclusion we are reaching is that in a customer environment, there will be direct competition with memory consumption during the execution of the instructions. And the degradation detected already in environments of other versions would possibly be due to Minions.

**Comment by Ch3LL** [comment]
> "after a certain time a degradation of salt-minions is detectable"

Do you mean that you could no longer communicate with them and get the time out error?

Just to be clear you did *NOT* seeing a memory leak correct? Maybe i'm misunderstanding here.

**Comment by erespaulo** [comment]
@Ch3LL ,

I'm conducting tests with Docker Containers to understand if the problem is with the minionswarm minion simulator.

But the behavior is that the minions, were not releasing the resources used after due processing, this way they died for "out of memory"

**Comment by erespaulo** [comment]
@Ch3LL ,

The tests with Containers have report a differnt answer cenario of minionswarm. 

In the tests with containers (3500 minions), I did not have instabilities in a continuous test of 48 hours.

I really couldn't come to any conclusions about the current 2019.2 environment;

The structure I uploaded will propose that these minions be migrated to this 3000.3 structure at first, and request the update of all minons in a gradual way for monitoring.
